builtin/umask.c

```diff
@@ -60,22 +186,26 @@ int builtin_umask(struct mrsh_state *state, int argc, char *argv[]) {
 	}
 
 	if (optind == argc) {
-		mode = umask(default_mode);
-		umask(mode);
+		mode = umask_current_mask();
+
 		if (umask_symbolic) {
 			umask_print_symbolic(mode);
 		} else {
 			printf("%04o\n", mode);
 		}
+
 		return EXIT_SUCCESS;
 	}
 
 	char *endptr;
 	mode = strtol(argv[optind], &endptr, 8);
+
 	if (*endptr != '\0') {
-		fprintf(stderr, "Invalid mode %s\n", argv[optind]);
-		fprintf(stderr, umask_usage);
-		return EXIT_FAILURE;
+		if (umask_mode(&mode, argv[optind]) != EXIT_SUCCESS) {
+			fprintf(stderr, "Invalid mode %s\n", argv[optind]);
```

Comment:
If the callee prints an error message, I don't think it's necessary to print another one here.